tests/h/auth/policy_test.py

```diff
@@ -109,6 +111,54 @@ def test_forget_uses_api_policy_for_api_paths(self, api_request):
         assert result == self.api_policy.forget.return_value
 
 
+class TestAuthClientAuthenticationPolicy(object):
+
+    def test_it_proxies_username_password_to_check_callback(self, pyramid_request, check):
+        check.return_value = []
+        policy = AuthClientPolicy(check=check)
+
+        policy.effective_principals(pyramid_request)
+
+        check.assert_called_once_with("foo", "bar", pyramid_request)
+
+    def test_it_returns_None_for_authenticated_userid(self, pyramid_request, check):
+        check.return_value = []
+        policy = AuthClientPolicy(check=check)
+
+        userid = policy.authenticated_userid(pyramid_request)
+
+        assert userid is None
+
+    def test_unauthenticated_userid_returns_client_id(self, pyramid_request, check):
+        check.return_value = []
+        policy = AuthClientPolicy(check=check)
+
+        userid = policy.unauthenticated_userid(pyramid_request)
+
+        assert userid == "foo"
+
+    def test_remember_does_nothing(self, pyramid_request, check):
+        policy = AuthClientPolicy(check=check)
+
+        assert policy.remember(pyramid_request, 'foo') == []
+
+    def test_forget_does_nothing(self, pyramid_request, check):
+        policy = AuthClientPolicy(check=check)
+
+        assert policy.forget(pyramid_request) == []
+
+    @pytest.fixture
+    def check(self):
+        return mock.Mock()
+
+    @pytest.fixture
+    def pyramid_request(self, pyramid_request):
```

Comment:
This method of doing this may be horrific. I honestly don't know.

Reply:
This seems right to me